File: infra/connectors/log-connector-python/smoke_test.py
#!/usr/bin/env python3
"""
smoke_test.py  --  LOG-003 integration smoke test

Verifies end-to-end log delivery from log_connector (Python) to:
  MongoDB  mongodb://127.0.0.1:27017/vibedev   collection: logs
  Elasticsearch  http://localhost:9200          index: kh-sim-YYYY.MM.DD

Exit codes:
  0  all checks passed
  1  one or more checks failed

Prerequisites:
  MongoDB 8.2.6 running on 27017 (DB-002 done)
  ELK stack up:  .\\_config\\Start-LocalEnv.ps1 -Action up -Stack elk
  pip install -r requirements.txt  (in this directory)

Run:
  python smoke_test.py
"""

from __future__ import annotations

import os
import sys
import time
import uuid
from datetime import datetime, timezone

# ── Configuration ─────────────────────────────────────────────────────────────

MONGO_URI    = os.environ.get("MONGODB_URI", "mongodb://127.0.0.1:27017")
ES_NODE      = os.environ.get("ES_NODE",     "http://localhost:9200")
SESSION      = f"smoke-{int(time.time())}"
APP          = "log-connector-smoke-test-py"
INDEX_PREFIX = "kh-sim"

# ── Utilities ─────────────────────────────────────────────────────────────────

passed = 0
failed = 0


def ok(label: str) -> None:
    global passed
    print(f"  [PASS] {label}")
    passed += 1


def fail(label: str, detail: str = "") -> None:
    global failed
    suffix = f": {detail}" if detail else ""
    print(f"  [FAIL] {label}{suffix}", file=sys.stderr)
    failed += 1


def header(text: str) -> None:
    print(f"\n{'─' * 60}")
    print(f"  {text}")
    print("─" * 60)


# ── Main ─────────────────────────────────────────────────────────────────────

def run() -> None:
    header("LOG-003 smoke test — log-connector-python")

    # ── 1. Emit test events ───────────────────────────────────────────────────
    header("Step 1: emit structured log events")

    try:
        from log_connector import get_logger
    except ImportError as exc:
        fail("Import log_connector", str(exc))
        _summary()
        return

    logger = get_logger(
        app=APP,
        session_id=SESSION,
        version="1.0.0",
        level="debug",
        mongo_uri=MONGO_URI,
        mongo_db="vibedev",
        mongo_collection="logs",
        es_node=ES_NODE,
        index_prefix=INDEX_PREFIX,
    )

    logger.info("Smoke test started", extra={"step": "init", "session": SESSION})
    logger.debug("Debug event", extra={"detail": "low-level trace"})
    logger.warning("Intentional warning", extra={"code": "W001", "threshold": 0.95})
    logger.error("Intentional error (non-fatal)", extra={"code": "E001", "recoverable": True})

    print("  Waiting 2s for handler flush...")
    time.sleep(2)

    ok("log events emitted (4 events: info, debug, warning, error)")

    # ── 2. Verify MongoDB ─────────────────────────────────────────────────────
    header("Step 2: verify MongoDB vibedev.logs")

    try:
        from pymongo import MongoClient
        from pymongo.errors import PyMongoError

        mc = MongoClient(MONGO_URI, serverSelectionTimeoutMS=5_000)
        mc.admin.command("ping")
        col   = mc["vibedev"]["logs"]
        count = col.count_documents({"session_id": SESSION, "app": APP})

        if count >= 4:
            ok(f"MongoDB: {count} documents found for session {SESSION}")
        elif count > 0:
            fail(f"MongoDB: expected >= 4 docs, found {count}")
        else:
            fail(f"MongoDB: 0 documents found for session {SESSION}")

        # Schema spot-check
        doc = col.find_one({"session_id": SESSION, "app": APP})
        if doc:
            required = all([
                doc.get("level"),
                doc.get("message"),
                doc.get("app"),
                doc.get("session_id"),
                isinstance(doc.get("metadata"), dict),
                doc["metadata"].get("host"),
                doc["metadata"].get("pid") is not None,
            ])
            if required:
                ok("MongoDB: document schema valid (level, message, app, session_id, metadata)")
            else:
                fail("MongoDB: document missing required schema fields", str(doc))

        mc.close()

    except ImportError:
        fail("MongoDB: pymongo not installed")
    except PyMongoError as exc:
        fail("MongoDB: connection or query failed", str(exc))

    # ── 3. Verify Elasticsearch ───────────────────────────────────────────────
    header("Step 3: verify Elasticsearch kh-sim-* index")

    try:
        from elasticsearch import Elasticsearch
        try:
            from elasticsearch.exceptions import ElasticsearchException
        except ImportError:
            from elasticsearch import ApiError as ElasticsearchException  # type: ignore

        es    = Elasticsearch([ES_NODE], request_timeout=10)
        today = datetime.now(timezone.utc).strftime("%Y.%m.%d")
        index = f"{INDEX_PREFIX}-{today}"

        # Force refresh so newly indexed docs are visible
        try:
            es.indices.refresh(index=f"{INDEX_PREFIX}-*")
        except ElasticsearchException:
            pass  # index may not exist yet if ES sink failed -- handled below

        # Diagnostic: total doc count (mapping-agnostic)
        total = es.count(index=f"{INDEX_PREFIX}-*")
        print(f"  [INFO] Total docs in {INDEX_PREFIX}-*: {total['count']}")

        # ES 8 dynamic mapping: string fields → text + .keyword sub-field.
        # Use session_id.keyword for exact term match.
        try:
            result = es.count(
                index=f"{INDEX_PREFIX}-*",
                query={"term": {"session_id.keyword": SESSION}},
            )
        except ElasticsearchException:
            result = es.count(
                index=f"{INDEX_PREFIX}-*",
                query={"match_phrase": {"session_id": SESSION}},
            )
        count = result["count"]

        if count >= 4:
            ok(f"Elasticsearch: {count} documents in {index} for session {SESSION}")
        elif count > 0:
            fail(f"Elasticsearch: expected >= 4 docs, found {count}")
        else:
            fail(f"Elasticsearch: 0 documents found in {index} for session {SESSION}")

        # Schema spot-check
        try:
            hit = es.search(
                index=f"{INDEX_PREFIX}-*",
                size=1,
                query={"term": {"session_id.keyword": SESSION}},
            )
        except ElasticsearchException:
            hit = {}
        hits = hit.get("hits", {}).get("hits", [])
        if hits:
            src = hits[0]["_source"]
            if src.get("@timestamp") and src.get("level"):
                ok("Elasticsearch: document has @timestamp and level fields")
            else:
                fail("Elasticsearch: document missing @timestamp or level", str(src))

    except ImportError:
        fail("Elasticsearch: elasticsearch-py not installed")
    except ElasticsearchException as exc:
        fail("Elasticsearch: connection or query failed", str(exc))

    # ── 4. Cluster health sanity ──────────────────────────────────────────────
    header("Step 4: ES cluster health")

    try:
        from elasticsearch import Elasticsearch
        es     = Elasticsearch([ES_NODE], request_timeout=5)
        health = es.cluster.health()
        status = health.get("status", "unknown")
        nodes  = health.get("number_of_nodes", "?")
        if status in ("green", "yellow"):
            ok(f"Elasticsearch cluster: status={status}, nodes={nodes}")
        else:
            fail(f"Elasticsearch cluster: unexpected status={status}")
    except Exception as exc:  # noqa: BLE001
        fail("Elasticsearch cluster health check failed", str(exc))

    # ── Summary ───────────────────────────────────────────────────────────────
    _summary()


def _summary() -> None:
    global passed, failed
    header("Summary")
    print(f"  Passed: {passed}")
    print(f"  Failed: {failed}")
    if failed == 0:
        print("\n  LOG-003 smoke test PASSED -- log-connector-python operational\n")
        sys.exit(0)
    else:
        print("\n  LOG-003 smoke test FAILED -- see [FAIL] lines above\n", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    run()

File: infra/connectors/log-connector-node/smoke-test.js
/**
 * smoke-test.js  --  LOG-002 integration smoke test
 *
 * Verifies end-to-end log delivery from log-connector-node to:
 *   MongoDB  mongodb://127.0.0.1:27017/vibedev  collection: logs
 *   Elasticsearch  http://localhost:9200         index: kh-sim-YYYY.MM.DD
 *
 * Exit codes:
 *   0  all checks passed
 *   1  one or more checks failed
 *
 * Prerequisites:
 *   MongoDB 27017 running (DB-002 done)
 *   ELK stack up: .\_config\Start-LocalEnv.ps1 -Action up -Stack elk
 *   npm install (in this directory)
 *
 * Run:
 *   node smoke-test.js
 *   npm run smoke-test
 */

'use strict';

const { MongoClient } = require('mongodb');
const { Client: ESClient } = require('@elastic/elasticsearch');
const { createLogger } = require('./index');

const MONGO_URI = process.env.MONGODB_URI || 'mongodb://127.0.0.1:27017';
const ES_NODE   = process.env.ES_NODE     || 'http://localhost:9200';
const SESSION   = `smoke-${Date.now()}`;
const APP       = 'log-connector-smoke-test';

// ── Utilities ─────────────────────────────────────────────────────────────────

let passed = 0;
let failed = 0;

function ok(label)   { console.log(`  [PASS] ${label}`); passed++; }
function fail(label, detail) {
  console.error(`  [FAIL] ${label}${detail ? ': ' + detail : ''}`);
  failed++;
}

function header(text) {
  console.log(`\n${'─'.repeat(60)}`);
  console.log(`  ${text}`);
  console.log('─'.repeat(60));
}

function sleep(ms) { return new Promise(r => setTimeout(r, ms)); }

// ── Main ─────────────────────────────────────────────────────────────────────

async function run() {
  header('LOG-002 smoke test — log-connector-node');

  // ── 1. Emit test events ───────────────────────────────────────────────────
  header('Step 1: emit structured log events');

  const logger = createLogger({
    app:       APP,
    sessionId: SESSION,
    version:   '1.0.0',
    level:     'debug',
    mongo:   { uri: MONGO_URI, db: 'vibedev', collection: 'logs' },
    elastic: { node: ES_NODE, indexPrefix: 'kh-sim' },
  });

  logger.info('Smoke test started', { step: 'init', session: SESSION });
  logger.debug('Debug event', { detail: 'low-level trace' });
  logger.warn('Intentional warning', { code: 'W001', threshold: 0.95 });
  logger.error('Intentional error (non-fatal)', { code: 'E001', recoverable: true });

  // Allow transports time to flush
  console.log('  Waiting 3s for transport flush...');
  await sleep(3000);

  // Explicitly close Mongo connection (Winston does not auto-close)
  for (const t of logger.transports) {
    if (typeof t.close === 'function') {
      await new Promise(res => t.close(res));
    }
  }

  ok('log events emitted (4 events: info, debug, warn, error)');

  // ── 2. Verify MongoDB ─────────────────────────────────────────────────────
  header('Step 2: verify MongoDB vibedev.logs');

  const mc = new MongoClient(MONGO_URI, { serverSelectionTimeoutMS: 5000 });
  try {
    await mc.connect();
    const col   = mc.db('vibedev').collection('logs');
    const count = await col.countDocuments({ session_id: SESSION, app: APP });

    if (count >= 4) {
      ok(`MongoDB: ${count} documents found for session ${SESSION}`);
    } else if (count > 0) {
      fail(`MongoDB: expected >= 4 docs, found ${count}`);
    } else {
      fail(`MongoDB: 0 documents found for session ${SESSION}`);
    }

    // Spot-check schema of one document
    const doc = await col.findOne({ session_id: SESSION, app: APP });
    if (doc) {
      const hasRequired = doc.level && doc.message && doc.app && doc.session_id &&
                          doc.metadata && doc.metadata.host && doc.metadata.pid;
      if (hasRequired) {
        ok('MongoDB: document schema valid (level, message, app, session_id, metadata.host, metadata.pid)');
      } else {
        fail('MongoDB: document missing required schema fields', JSON.stringify(doc));
      }
    }
  } catch (err) {
    fail('MongoDB: connection or query failed', err.message);
  } finally {
    await mc.close();
  }

  // ── 3. Verify Elasticsearch ───────────────────────────────────────────────
  header('Step 3: verify Elasticsearch kh-sim-* index');

  const es    = new ESClient({ node: ES_NODE, requestTimeout: 5000 });
  const today = new Date().toISOString().slice(0, 10).replace(/-/g, '.');
  const index = `kh-sim-${today}`;

  try {
    // Force index refresh so docs are visible to search immediately
    await es.indices.refresh({ index: `kh-sim-*` }).catch(() => {});

    // Diagnostic: total doc count in index (mapping-agnostic)
    const total = await es.count({ index: `kh-sim-*` });
    console.log(`  [INFO] Total docs in kh-sim-*: ${total.count}`);

    // ES 8 dynamic mapping: string fields → text + .keyword sub-field.
    // Use session_id.keyword for exact term match; fall back to match_phrase
    // against the analyzed text field if the index has no keyword sub-field.
    const result = await es.count({
      index:  `kh-sim-*`,
      query:  { term: { 'session_id.keyword': SESSION } },
    }).catch(() =>
      es.count({
        index: `kh-sim-*`,
        query: { match_phrase: { session_id: SESSION } },
      })
    );

    const count = result.count;
    if (count >= 4) {
      ok(`Elasticsearch: ${count} documents in ${index} for session ${SESSION}`);
    } else if (count > 0) {
      fail(`Elasticsearch: expected >= 4 docs, found ${count}`);
    } else {
      fail(`Elasticsearch: 0 documents found in ${index} for session ${SESSION}`);
    }

    // Spot-check mapping fields
    const hit = await es.search({
      index:  `kh-sim-*`,
      size:   1,
      query:  { term: { 'session_id.keyword': SESSION } },
    }).catch(() => null);

    if (hit && hit.hits.hits.length > 0) {
      const src = hit.hits.hits[0]._source;
      const hasTs = !!src['@timestamp'];
      const hasLvl = !!src.level;
      if (hasTs && hasLvl) {
        ok('Elasticsearch: document has @timestamp and level fields');
      } else {
        fail('Elasticsearch: document missing @timestamp or level', JSON.stringify(src));
      }
    }
  } catch (err) {
    fail('Elasticsearch: connection or query failed', err.message);
  }

  // ── 4. Cluster health sanity ──────────────────────────────────────────────
  header('Step 4: ES cluster health');
  try {
    const health = await es.cluster.health();
    if (health.status === 'green' || health.status === 'yellow') {
      ok(`Elasticsearch cluster: status=${health.status}, nodes=${health.number_of_nodes}`);
    } else {
      fail(`Elasticsearch cluster: unexpected status=${health.status}`);
    }
  } catch (err) {
    fail('Elasticsearch cluster health check failed', err.message);
  }

  // ── Summary ───────────────────────────────────────────────────────────────
  header('Summary');
  console.log(`  Passed: ${passed}`);
  console.log(`  Failed: ${failed}`);
  if (failed === 0) {
    console.log('\n  LOG-002 smoke test PASSED -- log-connector-node operational\n');
    process.exit(0);
  } else {
    console.error('\n  LOG-002 smoke test FAILED -- see [FAIL] lines above\n');
    process.exit(1);
  }
}

run().catch(err => {
  console.error('[smoke-test] Unhandled error:', err);
  process.exit(1);
});
